### PR TITLE
Villager state machine framework + implementation of a few states

### PR DESCRIPTION
**Pull request by bwrsandman**
This PR will probably be split up into more logical changes once it has been prototyped and tested.

TODO:
- [x] #384
- [ ] Make more ECS?

### CODE REVIEW

**Comment by raffclar** [comment]
I know this is WIP so no rush but I'm getting this

```
D:\openblack\src\Entities/Components/Villager.h(240): warning C4458: declaration of 'tribe' hides class member
D:\openblack\src\Entities/Components/Villager.h(284): note: see declaration of 'openblack::entities::components::Villager::tribe'
D:\openblack\src\Entities/Components/Villager.h(242): warning C4458: declaration of 'lifeStage' hides class member
D:\openblack\src\Entities/Components/Villager.h(282): note: see declaration of 'openblack::entities::components::Villager::lifeStage'
D:\openblack\src\Entities/Components/Villager.h(243): warning C4458: declaration of 'role' hides class member
D:\openblack\src\Entities/Components/Villager.h(285): note: see declaration of 'openblack::entities::components::Villager::role'
D:\openblack\src\Entities/Components/Villager.h(244): error C2064: term does not evaluate to a function taking 1 arguments
D:\openblack\src\Entities/Components/Villager.h(245): error C2064: term does not evaluate to a function taking 1 arguments
D:\openblack\src\Entities/Components/Villager.h(246): error C2064: term does not evaluate to a function taking 1 arguments
D:\openblack\src\Entities/Components/Villager.h(254): error C3536: 'h1': cannot be used before it is initialized
D:\openblack\src\Entities/Components/Villager.h(254): error C3536: 'h2': cannot be used before it is initialized
D:\openblack\src\Entities/Components/Villager.h(254): error C3536: 'h3': cannot be used before it is initialized
```

File: src/Entities/Components/Fixed.cpp [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/*****************************************************************************
+ * Copyright (c) 2018-2021 openblack developers
```
Comment (bwrsandman) [inline src/Entities/Components/Fixed.cpp:2]: 2022

File: src/Entities/Components/Fixed.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
+/*****************************************************************************
+ * Copyright (c) 2018-2021 openblack developers
```
Comment (bwrsandman) [inline src/Entities/Components/Fixed.h:2]: 2022

File: src/Entities/Components/Footpath.cpp [hunk 3]
```diff
@@ -0,0 +1,163 @@
+/*****************************************************************************
+ * Copyright (c) 2018-2021 openblack developers
```
Comment (bwrsandman) [inline src/Entities/Components/Footpath.cpp:2]: 2022

File: src/Entities/Components/Mobile.cpp [hunk 4]
```diff
@@ -0,0 +1,18 @@
+/*****************************************************************************
+ * Copyright (c) 2018-2021 openblack developers
```
Comment (bwrsandman) [inline src/Entities/Components/Mobile.cpp:2]: 2022

File: src/Entities/Components/Town.cpp [hunk 5]
```diff
@@ -0,0 +1,330 @@
+/*****************************************************************************
+ * Copyright (c) 2018-2021 openblack developers
```
Comment (bwrsandman) [inline src/Entities/Components/Town.cpp:2]: 2022

File: src/Entities/Components/Transform.cpp [hunk 6]
```diff
@@ -0,0 +1,45 @@
+/*****************************************************************************
+ * Copyright (c) 2018-2021 openblack developers
```
Comment (bwrsandman) [inline src/Entities/Components/Transform.cpp:2]: 2022

File: src/Entities/Map.cpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
+/*****************************************************************************
+ * Copyright (c) 2018-2021 openblack developers
```
Comment (bwrsandman) [inline src/Entities/Map.cpp:2]: 2022

File: src/Entities/Map.h [hunk 8]
```diff
@@ -0,0 +1,44 @@
+/*****************************************************************************
+ * Copyright (c) 2018-2021 openblack developers
```
Comment (bwrsandman) [inline src/Entities/Map.h:2]: 2022

File: src/InfoConstants.cpp [hunk 9]
```diff
@@ -0,0 +1,89 @@
+/*****************************************************************************
+ * Copyright (c) 2018-2021 openblack developers
```
Comment (bwrsandman) [inline src/InfoConstants.cpp:2]: 2022

File: src/Entities/Components/Abode.cpp [hunk 10]
```diff
@@ -0,0 +1,223 @@
+/*****************************************************************************
+ * Copyright (c) 2018-2021 openblack developers
```
Comment (bwrsandman) [inline src/Entities/Components/Abode.cpp:2]: 2022

File: src/Common/Spiral.cpp [hunk 11]
```diff
@@ -0,0 +1,61 @@
+/*****************************************************************************
+ * Copyright (c) 2018-2021 openblack developers
```
Comment (bwrsandman) [inline src/Common/Spiral.cpp:2]: 2022

**Comment by bwrsandman** [comment]
I dropped changes in the WIP commit but here it is for reference: https://github.com/openblack/openblack/pull/285/commits/cf319523

**Comment by bwrsandman** [comment]
Also noticed through this PR, a lot of logging is done without the macro

File: src/ECS/Components/Field.h [hunk 12]
```diff
@@ -14,7 +14,9 @@ namespace openblack::ecs::components
 
 struct Field
 {
-	int town;
+	static constexpr float radius2D = 5.0f;
```
Comment (bwrsandman) [inline src/ECS/Components/Field.h:17]: Can we get this from info constants?

File: src/ECS/Systems/LivingActionSystem.cpp [hunk 13]
```diff
@@ -0,0 +1,512 @@
...
+#include "Enums.h"
+#include "Game.h"
+
+using namespace openblack;
+using namespace openblack::ecs::components;
+using namespace openblack::ecs::systems;
+
+uint32_t VillagerInvalidState(LivingAction& action)
+{
+	SPDLOG_LOGGER_ERROR(spdlog::get("game"), "Villager #{}: Stuck in an invalid state",
```
Comment (bwrsandman) [inline src/ECS/Systems/LivingActionSystem.cpp:26]: Add a new logger called "ai"

File: src/ECS/Components/Abode.h [hunk 14]
```diff
@@ -18,6 +18,9 @@ namespace openblack::ecs::components
 
 struct Abode
 {
+	// FIXME(bwrsandman): Functions on components are not good
+	bool IsFunctional() const { return true; }
```
Comment (bwrsandman) [inline src/ECS/Components/Abode.h:22]: Move logic to manager

File: src/ECS/Archetypes/VillagerArchetype.cpp [hunk 15]
```diff
@@ -30,6 +33,11 @@ using namespace openblack::ecs::systems;
 entt::entity VillagerArchetype::Create([[maybe_unused]] const glm::vec3& abodePosition, const glm::vec3& position,
                                        VillagerInfo type, uint32_t age)
 {
+	// TODO(bwrsandman): would be nice to have this be a non-static member of
+	//  FeatureScriptCommands or use a singleton
+	thread_local std::default_random_engine generator;
+	thread_local std::uniform_int_distribution<uint16_t> duration_distribution(1, 500);
```
Comment (bwrsandman) [inline src/ECS/Archetypes/VillagerArchetype.cpp:39]: Add a random class that allows setting predictable seeds for testing

File: src/ECS/Components/Town.cpp [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+				// {
+				// 	this->physical->creature_3d->field_0x5228;
+				// }
+
+				// Object.Get2DRadius
+				else
+				{
+					bool any = registry.AnyOf<Fixed, Mobile>(entity);
+					assert(any);
+					assert(false); // TODO
```
Comment (bwrsandman) [inline src/ECS/Components/Town.cpp:99]: Unfinished code and not clear

File: src/ECS/Components/Town.cpp [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+	registry.Each<const Abode>([this, &checkPlanned](auto abode) {
+		if (abode.townId == id)
+		{
+			checkPlanned = false;
+		}
+	});
+
+	std::array<entt::entity, 100> fieldsAndPlanned;
+	uint32_t fieldsAndPlannedCount = 0; // TODO: rename to something about circular buffer
+	uint32_t abodeCount = 0;            // TODO: Rename to fieldsAndPlannedTotal
```
Comment (bwrsandman) [inline src/ECS/Components/Town.cpp:163]: Rename?

File: src/ECS/Map.cpp [hunk 18]
```diff
@@ -43,6 +43,17 @@ glm::vec2 Map::GetCellCenter(const Map::CellId& cellId)
...
 }
 
+glm::vec3 Map::PointGridCellAdd(const glm::vec3& pos, const glm::u16vec2& cellIncrement)
+{
+	const glm::u32vec2 shiftedIncrement(cellIncrement.x << 0x10, cellIncrement.y << 0x10);
+	const auto coords = glm::u32vec2(glm::xz(pos) * static_cast<float>(0x1000) / 10.0f) + shiftedIncrement;
+	return {
+	    coords.x * 10.0f / static_cast<float>(0x1000),
+	    pos.y,
+	    coords.y * 10.0f / static_cast<float>(0x1000),
```
Comment (bwrsandman) [inline src/ECS/Map.cpp:53]: Should be `/ 0x10000`

File: src/ECS/Systems/LivingActionSystem.cpp [hunk 19]
```diff
@@ -0,0 +1,519 @@
...
+	                       VillagerStateStrings[static_cast<size_t>(
+	                           LivingActionSystem::instance().VillagerGetState(action, LivingAction::Index::Top))]);
+	    return false;
+    },
+    .transitionAnimation = [](LivingAction& action) -> bool {
+	    SPDLOG_LOGGER_WARN(spdlog::get("ai"), "Villager #{}: TODO: Unimplemented transition animation function: {}",
+	                       static_cast<uint32_t>(Game::instance()->GetEntityRegistry().ToEntity(action)),
+	                       VillagerStateStrings[static_cast<size_t>(
+	                           LivingActionSystem::instance().VillagerGetState(action, LivingAction::Index::Top))]);
+	    return -1;
```
Comment (bwrsandman) [inline src/ECS/Systems/LivingActionSystem.cpp:111]: `return false;`

File: src/ECS/Systems/LivingActionSystem.cpp [hunk 20]
```diff
@@ -0,0 +1,519 @@
...
+	                       VillagerStateStrings[static_cast<size_t>(
+	                           LivingActionSystem::instance().VillagerGetState(action, LivingAction::Index::Top))]);
+	    return false;
+    },
+};
+
+std::array<VillagerStateTableEntry, static_cast<size_t>(VillagerStates::_COUNT)> VillagerStateTable = {
+    /* INVALID_STATE */ VillagerStateTableEntry {
+        .state = &VillagerInvalidState,
+    },
```
Comment (bwrsandman) [inline src/ECS/Systems/LivingActionSystem.cpp:125]: Add missing initializers

File: src/ECS/Systems/LivingActionSystem.cpp [hunk 21]
```diff
@@ -0,0 +1,519 @@
...
+	return static_cast<VillagerStates>(action.states[static_cast<size_t>(index)]);
+}
+
+void LivingActionSystem::VillagerSetState(LivingAction& action, LivingAction::Index index, VillagerStates state,
+                                          bool skipTransition) const
+{
+	assert(index < LivingAction::Index::_Count);
+	if (static_cast<VillagerStates>(action.states[static_cast<size_t>(index)]) != state)
+	{
+		auto& registry = Game::instance()->GetEntityRegistry();
```
Comment (bwrsandman) [inline src/ECS/Systems/LivingActionSystem.cpp:435]: Remove this line

**Comment by raffclar** [comment]
About making this more "ECS". I think we should explore the way we update our systems per tick - but not in this PR. EnTT provides a minimal robust cooperative scheduler with lots of neat quirks like a graceful abort. Go to https://github.com/skypjack/entt/wiki/Crash-Course:-cooperative-scheduler and look at the "chains of processes" paragraph and you can see how we might be able to chain together our systems if a fairely logical way

**Review by raffclar — CHANGES_REQUESTED** [review]

**Comment by raffclar** [comment]
The interactive debug gui is amazing by the way

File: src/Enums.h [hunk 22]
```diff
@@ -1301,8 +1301,270 @@ enum class VillagerStates : uint32_t
 	WaitForArtifactDance = 253,
 	BreederJustLanded = 254,
 	LastState = 255,
+
+	_COUNT
 };
 
+static constexpr std::array<std::string_view, static_cast<size_t>(VillagerStates::_COUNT)> VillagerStateStrings = {
```
[the remark this answers is not in the document]
Comment (bwrsandman) [reply src/Enums.h:1308]: There should be 256 `VillagerStates` not including `_COUNT`. The first one is a 0.

[the remark this answers is not in the document]
Comment (bwrsandman) [reply src/Enums.h:1308]: Good catch

File: src/ECS/Systems/LivingActionSystem.cpp [hunk 23]
```diff
@@ -0,0 +1,519 @@
...
+}
+
+void LivingActionSystem::VillagerSetState(LivingAction& action, LivingAction::Index index, VillagerStates state,
+                                          bool skipTransition) const
+{
+	assert(index < LivingAction::Index::_Count);
+	if (static_cast<VillagerStates>(action.states[static_cast<size_t>(index)]) != state)
+	{
+		[[maybe_unused]] auto& registry = Game::instance()->GetEntityRegistry();
+		SPDLOG_LOGGER_DEBUG(spdlog::get("ai"), "Villager #{}: Setting state {} -> {}",
```
[the remark this answers is not in the document]
Comment (bwrsandman) [reply src/ECS/Systems/LivingActionSystem.cpp:436]: If I set it to trace is that good enough?

[the remark this answers is not in the document]
Comment (raffclar) [reply src/ECS/Systems/LivingActionSystem.cpp]: Yeah trace is better

**Comment by raffclar** [comment]
There's an infinite loop occurring in the fitBox function. Digging a bit deeper into now.

**Review by raffclar — APPROVED** [review]

**Pull request merged**